test/sshkit_functional_test.exs

```diff
@@ -17,6 +17,22 @@ defmodule SSHKitFunctionalTest do
       assert name == host.options[:user]
     end
 
+    @tag boot: [@bootconf]
+    test "connects as the login user and runs commands in parallel", %{hosts: [host]} do
+      begin_time = Time.utc_now()
+      [{:ok, output1, 0},{:ok, output2, 0}] =
+        [host, host]
+        |> SSHKit.context()
+        |> SSHKit.run("sleep 2; id -un", :parallel)
+      end_time = Time.utc_now()
+      run_time = Time.diff(end_time, begin_time, :second)
+
+      assert run_time < 4
+      assert String.trim(stdout(output1)) == host.options[:user]
+      assert String.trim(stdout(output2)) == host.options[:user]
+
+    end
```

Comment:
@tessi do you have an idea?

Reply:
or @klappradla?

Reply:
We might need two tests :)

1. to test parallel execution on two different hosts (as @pmeinhardt  said) which is not needed to test parallelism and instead focuses on correct execution (so we see when things break for real users in a real'ish scenario).
2. test parallel execution on the same host, in this time also testing that things run in parallel.

For 2. I agree that the time approach might work -- but it makes our test suite at least 2 seconds slower. Instead I would suggest to test a side effect on that host which can only happen when two parallel ssh sessions happen. The question is which side effect to test. Possible options are:

* monitor currently active ssh shells. The command could monitor active ssh sessions and wait until two sessions are seen (maybe with a timeout to prevent lifelock).
* kinda crappy: a command could append numbers from 1 to 1.000.000 into a file. If both ssh sessions do that the file content would contain "unsorted" numbers
* if the two commands could be different: one could open a socket/file and read/wait on it. the other could write to that socket/file.

I personally prefer the first option, as it's the least crappy one :)
It seems the `who` command lists currently logged in users

> The who utility displays a list of all users currently logged on, showing for each user the login name, tty name, the date and time of login, and hostname if not local.

We could write (and then execute) a little bash script which runs `who` until two users are logged in (again, a timeout might be nice). That bash script could be a while-loop wrapped in the [`timeout`](https://ss64.com/bash/timeout.html) command.

Reply:
Oh and again (because I forgot to write this before), thanks @seungjin for your contribution 💚 I would love this feature to go in.